core/src/execution/datafusion/expressions/binary.rs

```diff
@@ -0,0 +1,211 @@
+            Arc::new(BinaryExpr::new(
+                Arc::new(BinaryExpr::new(
+                    self.left.clone(),
+                    Operator::BitwiseXor,
+                    self.inner.clone(),
+                )),
+                Operator::BitwiseAnd,
+                Arc::new(BinaryExpr::new(
+                    self.right.clone(),
+                    Operator::BitwiseXor,
+                    self.inner.clone(),
+                )),
+            )),
+            Operator::Lt,
+            Self::zero_literal(&result.data_type())?,
+        ));
+        match check_overflow_expr.evaluate(batch)? {
+            ColumnarValue::Array(array) => {
+                let boolean_array = array.as_any().downcast_ref::<BooleanArray>();
+                match boolean_array {
+                    Some(boolean_array) => {
+                        if boolean_array.true_count() > 0 {
+                            return Err(
+                                arithmetic_overflow_error(&result.data_type().to_string()).into()
+                            );
+                        }
+                    }
+                    None => {
+                        return Err(DataFusionError::Internal(format!(
+                            "Expected boolean array, found: {:?}",
+                            array
+                        )));
+                    }
+                }
+                Ok(())
+            }
+            ColumnarValue::Scalar(ScalarValue::Boolean(Some(true))) => {
+                Err(arithmetic_overflow_error(&result.data_type().to_string()).into())
+            }
+            _ => Ok(()),
+        }
+    }
+
+    fn zero_literal(data_type: &DataType) -> Result<Arc<dyn PhysicalExpr>> {
+        let zero_literal = match data_type {
+            DataType::Int8 => ScalarValue::Int8(Some(0)),
+            DataType::Int16 => ScalarValue::Int16(Some(0)),
+            DataType::Int32 => ScalarValue::Int32(Some(0)),
+            DataType::Int64 => ScalarValue::Int64(Some(0)),
+            _ => {
+                return Err(DataFusionError::Internal(format!(
+                    "Unsupported data type: {:?}",
+                    data_type
+                )))
+            }
+        };
+        Ok(Arc::new(Literal::new(zero_literal)))
+    }
+}
+
+impl Display for CometBinaryExpr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.inner.fmt(f)
+    }
+}
+
+impl PhysicalExpr for CometBinaryExpr {
+    fn as_any(&self) -> &dyn Any {
+        self.inner.as_any()
+    }
+
+    fn data_type(&self, input_schema: &Schema) -> Result<DataType> {
+        self.inner.data_type(input_schema)
+    }
+
+    fn nullable(&self, input_schema: &Schema) -> Result<bool> {
+        self.inner.nullable(input_schema)
+    }
+
+    fn evaluate(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
+        match self.inner.evaluate(batch) {
+            Ok(result) => {
+                self.fail_on_overflow(batch, &result)?;
```

Comment:
Evaluating the same expression twice is going to be expensive. We should just evaluate once, either checking for overflows, or not, depending on eval mode.